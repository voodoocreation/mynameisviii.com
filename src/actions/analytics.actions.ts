import actionCreatorFactory from "typescript-fsa";

const createAction = actionCreatorFactory("ANALYTICS");

export const trackEvent = createAction<{
  event: string;
  [index: string]: any;
}>("TRACK_EVENT");
