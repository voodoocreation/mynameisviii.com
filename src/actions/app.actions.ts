import actionCreatorFactory from "typescript-fsa";

const createAction = actionCreatorFactory("APP");

export const initApp = createAction.async<{ locale?: string }, {}>("INIT");

export const toggleNavigation = createAction("TOGGLE_NAVIGATION");

export const setOnlineStatus = createAction<boolean>("SET_ONLINE_STATUS");

export const setHasNewVersion = createAction<boolean>("SET_HAS_NEW_VERSION");

export const setCurrentRoute = createAction<string>("SET_CURRENT_ROUTE");

export const changeRoute = createAction.async<string, {}>("CHANGE_ROUTE");
