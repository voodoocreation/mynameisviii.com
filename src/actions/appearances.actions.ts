import actionCreatorFactory from "typescript-fsa";

import { IAppearance, IDynamoResponse, ILatLng } from "../models/root.models";

const createAction = actionCreatorFactory("APPEARANCES");

export const fetchAppearances = createAction.async<
  {},
  IDynamoResponse<IAppearance, "startingAt">
>("FETCH");

export const fetchMoreAppearances = createAction.async<
  {},
  IDynamoResponse<IAppearance, "startingAt">
>("FETCH_MORE");

export const fetchAppearanceBySlug = createAction.async<string, IAppearance>(
  "FETCH_BY_SLUG"
);

export const setCurrentAppearanceSlug = createAction<string>(
  "SET_CURRENT_SLUG"
);

export const geocodeCurrentAppearanceAddress = createAction.async<{}, ILatLng>(
  "GEOCODE_CURRENT_ADDRESS"
);
