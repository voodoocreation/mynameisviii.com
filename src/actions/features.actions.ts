import actionCreatorFactory from "typescript-fsa";

const createAction = actionCreatorFactory("FEATURES");

export const addFeatures = createAction<string | string[]>("ADD");

export const removeFeatures = createAction<string | string[]>("REMOVE");
