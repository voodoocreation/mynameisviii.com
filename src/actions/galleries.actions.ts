import actionCreatorFactory from "typescript-fsa";

import { IGallery, IS3Response } from "../models/root.models";

const createAction = actionCreatorFactory("GALLERIES");

export const fetchGalleries = createAction.async<{}, IS3Response<IGallery>>(
  "FETCH"
);

export const fetchMoreGalleries = createAction.async<{}, IS3Response<IGallery>>(
  "FETCH_MORE"
);

export const fetchGalleryBySlug = createAction.async<string, IGallery>(
  "FETCH_BY_SLUG"
);

export const setCurrentGallerySlug = createAction<string>("SET_CURRENT_SLUG");
