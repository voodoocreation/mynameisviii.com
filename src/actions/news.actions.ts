import actionCreatorFactory from "typescript-fsa";

import { IDynamoResponse, INewsArticle } from "../models/root.models";

const createAction = actionCreatorFactory("NEWS");

export const fetchLatestNews = createAction.async<
  {},
  IDynamoResponse<INewsArticle, "createdAt">
>("FETCH");

export const fetchMoreLatestNews = createAction.async<
  {},
  IDynamoResponse<INewsArticle, "createdAt">
>("FETCH_MORE");

export const fetchNewsArticleBySlug = createAction.async<string, INewsArticle>(
  "FETCH_BY_SLUG"
);

export const setCurrentNewsArticleSlug = createAction<string>(
  "SET_CURRENT_SLUG"
);
