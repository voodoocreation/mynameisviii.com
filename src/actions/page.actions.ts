import actionCreatorFactory from "typescript-fsa";

const actionCreator = actionCreatorFactory("PAGE");

export const toggleNavigation = actionCreator<{}>("TOGGLE_NAVIGATION");

export const changeRoute = actionCreator.async<
  PLChangeRouteStarted,
  PLChangeRouteDone,
  PLChangeRouteFailed
>("CHANGE_ROUTE");
