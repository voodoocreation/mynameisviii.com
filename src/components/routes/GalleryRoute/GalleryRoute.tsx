import Head from "next/head";
import * as React from "react";
import { InjectedIntlProps } from "react-intl";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";

import { absUrl } from "../../../helpers/dataTransformers";
import injectIntlIntoPage from "../../../helpers/injectIntlIntoPage";
import { IGallery } from "../../../models/root.models";
import { TStoreState } from "../../../reducers/root.reducers";
import { IPageContext } from "../../connected/App/App";
import Gallery from "../../presentation/Gallery/Gallery";
import Loader from "../../presentation/Loader/Loader";

import * as actions from "../../../actions/root.actions";
import * as selectors from "../../../selectors/root.selectors";

interface IProps extends InjectedIntlProps {
  fetchGalleryBySlug: typeof actions.fetchGalleryBySlug.started;
  gallery?: IGallery;
  isLoading: boolean;
  trackEvent: typeof actions.trackEvent;
}

class GalleryRoute extends React.Component<IProps> {
  public static async getInitialProps(context: IPageContext) {
    const { query, store } = context;
    const slug = query.slug as string;

    store.dispatch(actions.setCurrentGallerySlug(slug));

    const state = store.getState();
    const currentGallery = selectors.getCurrentGallery(state);

    if (!currentGallery || !currentGallery.images) {
      store.dispatch(actions.fetchGalleryBySlug.started(slug));
    }
  }

  public render() {
    const { gallery, isLoading } = this.props;
    const { formatMessage } = this.props.intl;

    if (isLoading) {
      return <Loader className="PageLoader" />;
    }

    if (!gallery) {
      return null;
    }

    return (
      <React.Fragment>
        <Head>
          <title>
            {gallery.title}
            {" Â· "}
            {formatMessage({ id: "BRAND_NAME" })}
          </title>

          <meta content={gallery.description} name="description" />

          <meta property="og:title" content={gallery.title} />
          <meta property="og:description" content={gallery.description} />
          <meta
            property="og:url"
            content={absUrl(`/galleries/${gallery.slug}`)}
          />
          <meta property="og:type" content="website" />
          <meta property="og:image" content={gallery.imageUrl} />
        </Head>

        <Gallery
          {...gallery}
          onGalleryInteraction={this.onGalleryInteraction}
        />
      </React.Fragment>
    );
  }

  private onGalleryInteraction = (type: string, index?: number) => {
    this.props.trackEvent({
      event: `gallery.gallery.${type}`,
      index
    });
  };
}

const mapState = (state: TStoreState) => ({
  gallery: selectors.getCurrentGallery(state),
  isLoading: selectors.getGalleriesIsLoading(state)
});

const mapActions = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      fetchGalleryBySlug: actions.fetchGalleryBySlug.started,
      trackEvent: actions.trackEvent
    },
    dispatch
  );

export default injectIntlIntoPage(
  connect(
    mapState,
    mapActions
  )(GalleryRoute)
);
