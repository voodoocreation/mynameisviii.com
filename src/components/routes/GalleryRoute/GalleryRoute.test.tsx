import * as actions from "../../../actions/root.actions";
import { gallery } from "../../../models/root.models";
import ComponentTester from "../../../utilities/ComponentTester";
import MockPageContext from "../../../utilities/MockPageContext";
import GalleryRoute from "./GalleryRoute";

const item = gallery({
  images: [{ imageUrl: "1" }, { imageUrl: "2" }],
  slug: "test-1"
});

const defaultState = {
  galleries: {
    currentSlug: item.slug,
    items: {
      [item.slug]: item
    }
  }
};

const component = new ComponentTester(GalleryRoute, true);

describe("[routes] <GalleryRoute />", () => {
  describe("getInitialProps", () => {
    const context = new MockPageContext()
      .withDefaultQuery({
        slug: item.slug
      })
      .withDefaultReduxState(defaultState);

    describe("when the gallery isn't already in the store", () => {
      const slug = "new-slug";

      it("calls getInitialProps method", async () => {
        await GalleryRoute.getInitialProps(
          context.withQuery({ slug }).toObject()
        );
      });

      it("dispatches actions.setCurrentGallerySlug with expected payload", () => {
        const matchingActions = context.reduxHistory.filter(
          actions.setCurrentGallerySlug.match
        );

        expect(matchingActions).toHaveLength(1);
        expect(matchingActions[0].payload).toBe(slug);
      });

      it("dispatches actions.fetchGalleryBySlug.started", () => {
        expect(
          context.reduxHistory.filter(actions.fetchGalleryBySlug.started.match)
        ).toHaveLength(1);
      });
    });

    describe("when the gallery is already in the store", () => {
      it("calls getInitialProps method", async () => {
        await GalleryRoute.getInitialProps(context.toObject());
      });

      it("dispatches actions.setCurrentGallerySlug", () => {
        expect(
          context.reduxHistory.filter(actions.setCurrentGallerySlug.match)
        ).toHaveLength(1);
      });

      it("doesn't dispatch actions.fetchGalleryBySlug.started", () => {
        expect(
          context.reduxHistory.filter(actions.fetchGalleryBySlug.started.match)
        ).toHaveLength(0);
      });
    });
  });

  it("matches snapshot", () => {
    const { wrapper } = component.withReduxState(defaultState).mount();

    expect(wrapper.render()).toMatchSnapshot();
  });

  it("renders a loader when isLoading is true", () => {
    const { wrapper } = component
      .withReduxState({
        ...defaultState,
        galleries: { isLoading: true }
      })
      .mount();

    expect(wrapper.find("Loader")).toHaveLength(1);
  });

  it("doesn't render anything when no gallery exists", () => {
    const { wrapper } = component
      .withReduxState({
        galleries: {
          currentSlug: undefined,
          hasError: true,
          items: {}
        }
      })
      .mount();

    expect(wrapper.render().html()).toBeNull();
  });

  it("tracks gallery interactions correctly", () => {
    const { wrapper } = component.withReduxState(defaultState).mount();

    wrapper
      .find("ImageGallery Image")
      .last()
      .simulate("click");

    const matchingActions = component.reduxHistory.filter(
      actions.trackEvent.match
    );

    expect(matchingActions).toHaveLength(1);
    expect(matchingActions[0].payload).toEqual({
      event: "gallery.gallery.itemClick",
      index: 1
    });
  });
});
