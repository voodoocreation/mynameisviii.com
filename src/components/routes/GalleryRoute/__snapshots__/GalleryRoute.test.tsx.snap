// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[routes] <GalleryRoute /> matches snapshot 1`] = `
<article
  class="Gallery"
>
  <header
    class="PageHeader"
  >
    <h1 />
  </header>
  <div
    class="Gallery--description"
  >
    <p />
  </div>
  <section
    class="Gallery--images"
  >
    <div
      class="ImageGallery"
    >
      <figure
        class="Image Gallery--image isLoading"
        role="button"
        tabindex="0"
      >
        <div
          class="Image--wrapper"
        >
          <img
            alt=" - 1"
            src="1"
          />
          <span
            class="Loader"
          >
            <div
              class="Loader--animation"
            >
              <span />
              <span />
              <span />
            </div>
          </span>
        </div>
      </figure>
      <figure
        class="Image Gallery--image isLoading"
        role="button"
        tabindex="0"
      >
        <div
          class="Image--wrapper"
        >
          <img
            alt=" - 2"
            src="2"
          />
          <span
            class="Loader"
          >
            <div
              class="Loader--animation"
            >
              <span />
              <span />
              <span />
            </div>
          </span>
        </div>
      </figure>
    </div>
  </section>
</article>
`;
